ambry-replication/src/main/java/com/github/ambry/replication/PartitionLeaderInfo.java

```diff
@@ -0,0 +1,163 @@
+    // 3. Log the list of leader replicas associated with this partition (will be used later for leadership based replication)
+    List<ReplicaId> peerLeaderReplicas = new ArrayList<>();
+    for (ReplicaId leaderReplica : leaderReplicas) {
+      if (leaderReplica.getDataNodeId() != localReplica.getDataNodeId()) {
+        peerLeaderReplicas.add(leaderReplica);
+        logger.info("Partition {} on node instance {} is leader in remote dc {}", partitionName,
+            getInstanceName(leaderReplica.getDataNodeId().getHostname(), leaderReplica.getDataNodeId().getPort()),
+            leaderReplica.getDataNodeId().getDatacenterName());
+      }
+    }
+
+    // Read-write lock avoids contention from threads removing old leader partitions (removePartition()) and threads updating existing leader partitions (refreshPeerLeadersForAllPartitions())
+    rwLock.writeLock().lock();
+    try {
+      peerLeaderReplicasByPartition.put(partitionName, new HashSet<>(peerLeaderReplicas));
+    } finally {
+      rwLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Remove a partition from the map of leader partitions. This method is thread safe.
+   * @param partitionName name of the partition to be removed
+   */
+  public void removePartition(String partitionName) {
+    // Read-write lock avoids contention from threads adding new leaders (addPartition()) and threads updating existing leader partitions (refreshPeerLeadersForAllPartitions())
+    rwLock.writeLock().lock();
+    try {
+      peerLeaderReplicasByPartition.remove(partitionName);
+    } finally {
+      rwLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Refreshes the list of remote leaders for all leader partitions (by looking at latest leader set in RoutingTableSnapshot). This method is thread safe.
+   */
+  public void refreshPeerLeadersForAllPartitions() {
+
+    // Read-write lock usage:
+    // Avoids contention with threads adding new leaders (in addPeerLeadersByPartition()) and removing old leaders (in removePartition()).
+    // Multiple threads can call this method in parallel as we only update existing partitions (no adding or removing). Since, it is a concurrent hash map, PUTs and GETs will be clean.
+
+    rwLock.readLock().lock();
+    try {
+      for (Map.Entry<String, Set<ReplicaId>> entry : peerLeaderReplicasByPartition.entrySet()) {
+        String partitionName = entry.getKey();
+        ReplicaId localLeaderReplica = storeManager.getReplica(partitionName);
+        PartitionId partition = localLeaderReplica.getPartitionId();
+        Set<ReplicaId> previousRemoteLeaderReplicas = entry.getValue();
+        Set<ReplicaId> currentRemoteLeaderReplicas =
+            new HashSet<>(partition.getReplicaIdsByState(ReplicaState.LEADER, null));
+        currentRemoteLeaderReplicas.remove(localLeaderReplica);
+        if (!previousRemoteLeaderReplicas.equals(currentRemoteLeaderReplicas)) {
+          peerLeaderReplicasByPartition.put(partitionName, currentRemoteLeaderReplicas);
```

Comment:
There will be race condition if we use read-lock here.  Following is an example(leader):
           remote DC1           remote DC2
T1:        R1(R2, R3)             R4(R5, R6)
T2:       R2 (R1, R3)             R4(R5, R6)           ---> current remote leader replica (R2, R4)
T3:       R2 (R1, R3)             R5(R4, R6).          ---> current remote leader replica (R2, R5) 

It's possible (R2, R5) get in first and then be overridden to (R2, R4)